src/java.desktop/macosx/native/libosxapp/JNIUtilities.h

```diff
@@ -147,9 +154,44 @@
 
 /*********       EXCEPTION_HANDLING    *********/
 
+/*
+ * Some explanation to set context of the bigger picture.
+ * Before returning to Java from JNI, NSExceptions are caught - so long as
+ * the body of the native method is wrapped in the ENTER/EXIT macros.
+ * So if we want to directly return to Java from some nested Objective-C
+ * function when detecting a Java exception, we just need to raise an
+ * NSException. Then clear that right before returning to Java,
+ * leaving the Java exception to be seen back in Java-land.
+ *
+ * But if the current thread is the Appkit thread we might as well clear
+ * the Java Exception right now since there's nothing to receive it.
+ * In such a case control will propagate back to the run loop which might
+ * terminate the application. One drawback of that is that the location of
+ * termination does not show where the NSException originated.
+ * And for whatever reason, something swallows that exception.
+ * So as a debugging aid, when on the AppKit thread we can provide a
+ * way (via an env. var.) to log the location.
+ * Additionally provide a similar way to prevent the NSException being
+ * raised and instead just clear the Java Exception.
+ * Together these provide alternate behaviours for more debugging info
+ * or maybe a way for the app to continue running depending on the exact
+ * nature of the problem that has been detected and how survivable it is.
+ */
 #define CHECK_EXCEPTION() \
```

Comment:
Since this macro does not try to return from the outer code we can change it to the method?

Reply:
But then "env" would need to be passed explicitly. I don't think the churn is worth it.

Reply:
But that could be merged to the CallXXXMethod and placed somewhere in the libosxapp

Reply:
You mean those methods  or macros that we already discussed and I explained why I do not want to do that ?
This is a small update for error checking not revisiting everything done before.